### ISSUE TITLE
Add support for grad calculation in validation/test loop

### ISSUE THREAD

**Issue by rhythmswing** [PyTorchLightning/pytorch-lightning#10287]
Hi,

It seems that the lightning framework automatically disables autograd in testing and validation loops. While this is generally desired in evaluation mode, there are cases when the differentiation is a part of the computation flow (e.g. in physics engines where you desire to calculate the derivatives of coordinates.) 

Would it be possible to add options to enable evaluation mode (as in model.eval()) but not forcibly disable autograd while in validation/testing loop? 

Thanks!

**Comment by rohitgr7** [comment]
you can enable it for your usecase:
```py
class LitModel(LightningModule):
    def on_validation_model_eval(self, *args, **kwargs):
        super().on_validation_model_eval(*args, **kwargs)
        torch.set_grad_enabled(True)
```
we have `on_validation_model_train`, `on_test_model_eval`, `on_test_model_train` too.

the sequence for validation is like:
```
1. Lightning disable autograd
2. on_validation_model_eval() is called
3. val loop runs
4. on_validation_model_train() is called
```
and similarly for test

**Comment by rhythmswing** [comment]
Hi, what is the purpose of on_validation_model_train and on_test_model_train? Whats the difference between them and _eval?

**Comment by rhythmswing** [comment]
> you can enable it for your usecase:
> 
> ```python
> class LitModel(LightningModule):
>     def on_validation_model_eval(self, *args, **kwargs):
>         super().on_validation_model_eval(*args, **kwargs)
>         torch.set_grad_enabled(True)
> ```
> 
> we have `on_validation_model_train`, `on_test_model_eval`, `on_test_model_train` too.
> 
> the sequence for validation is like:
> 
> ```
> 1. Lightning disable autograd
> 2. on_validation_model_eval() is called
> 3. val loop runs
> 4. on_validation_model_train() is called
> ```
> 
> and similarly for test

Thanks! 
What is the purpose of on_validation_model_train and on_test_model_train? Whats the difference between them and _eval?

**Comment by rohitgr7** [comment]
These hooks are called and added for user right before the corresponding loop starts. Before the validation loop starts we call _eval method to that internally just calls model.eval to disable dropout and batchnorm. _train methods are called called after these loops end to put back the model again in the training mode.

**Comment by rhythmswing** [comment]
Thanks.

**Comment by Zian-0427** [comment]
> These hooks are called and added for user right before the corresponding loop starts. Before the validation loop starts we call _eval method that internally just calls model.eval to disable dropout and batchnorm. _train methods are called called after these loops end to put back the model again in the training mode.



> you can enable it for your usecase:
> 
> ```python
> class LitModel(LightningModule):
>     def on_validation_model_eval(self, *args, **kwargs):
>         super().on_validation_model_eval(*args, **kwargs)
>         torch.set_grad_enabled(True)
> ```
> 
> we have `on_validation_model_train`, `on_test_model_eval`, `on_test_model_train` too.
> 
> the sequence for validation is like:
> 
> ```
> 1. Lightning disable autograd
> 2. on_validation_model_eval() is called
> 3. val loop runs
> 4. on_validation_model_train() is called
> ```
> 
> and similarly for test

Hi,

I've met the same problem and adopted your method to try to fix it. However, when I enabled grad in `on_test_model_eval ` like you and try to call `trainer.test(model, dataloader)`, the same problem arose and it doesn't seem to work. Why?

Thanks!

**Comment by rohitgr7** [comment]
hey @Zian-0427 !

https://github.com/Lightning-AI/lightning/issues/14497 will resolve your use-case

**Comment by Zian-0427** [comment]
> 

Thank you for replying! I've read #14497 and probably understand how to solve this problem. 

Does it mean that I need to add the following codes in my test_step to close the inference mode and make a normal tensor copy from the input data?

`with torch.inference_mode(False):
            origin_data = batch
            origin_data = origin_data.clone()
`

**Comment by Zian-0427** [comment]
Thank you for replying! I've read https://github.com/Lightning-AI/lightning/issues/14497 and probably understand how to solve this problem.

Does it mean that I need to add the following codes in my test_step to close the inference mode and make a normal tensor copy from the input data?
```python 
with torch.inference_mode(False): 
     origin_data = batch 
     origin_data = origin_data.clone()
     ...
